File: LeetCode/ArrayStrings/python/jumpGameII.py
from typing import List


class Solution:
    def jump(self, nums: List[int]) -> int:
        n = len(nums)
        res = 0
        l = 0
        r = 0

        while r < n-1:
            farthest = 0

            for i in range(l, r + 1):
                farthest = max(farthest, i + nums[i])
            
            l = r + 1
            r = farthest

            res += 1

        return res
    
    @staticmethod
    def testSolution(input):
        print("Input: nums: {}".format(input["nums"]))
        print("Expected: {}".format(input["expected"]))
        print("Result: {}".format(Solution().jump(input["nums"])))
        print("-" * 50)


if __name__ == "__main__":
    input = {"nums" : [2,3,1,1,4], "expected" : 2}
    Solution.testSolution(input)
    
    input = {"nums" : [2,3,0,1,4], "expected" : 2}
    Solution.testSolution(input)

File: LeetCode/ArrayStrings/java/jumpGameII.java
package LeetCode.ArrayStrings.java;

import java.util.Arrays;
import java.util.stream.IntStream;

record jumpGameIIRecord(int[] nums, int expected) {

}

public class jumpGameII {
    public int jump(int[] nums) {
        int n = nums.length;
        int res = 0;
        int l = 0;
        int r = 0;

        while (r < n - 1) {
            int farthest = 0;

            for (int i : IntStream.range(l, r + 1).toArray()) {
                farthest = Math.max(farthest, i + nums[i]);
            }

            l = r + 1;
            r = farthest;

            res += 1;
        }

        return res;
    }

    public static void main(String[] args) {
        jumpGameIIRecord input = new jumpGameIIRecord(new int[] { 2, 3, 1, 1, 4 }, 2);
        jumpGameII.testSolution(input);

        input = new jumpGameIIRecord(new int[] { 2, 3, 0, 1, 4 }, 2);
        jumpGameII.testSolution(input);
    }

    private static void testSolution(jumpGameIIRecord input) {
        System.out.println("Input: nums: " + Arrays.toString(input.nums()));
        System.out.println("Expected: " + input.expected());
        System.out.println("Result: " + new jumpGameII().jump(input.nums()));
        System.out.println("-".repeat(50));
    }
}
